src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs

```diff
@@ -3,311 +3,819 @@
+                item.SetMetadata(MetadataKeys.HintPath, item.ItemSpec);
+                item.SetMetadata(MetadataKeys.ExternallyResolved, externallyResolved);
+            }
+        }
 
-            CompileTimeAssemblies = RaisePackageAssets(
-                compileTimeTarget,
-                package => package.CompileTimeAssemblies,
-                setup: (asset, item) =>
-                {
-                    item.SetMetadata(MetadataKeys.Private, "false");
-                    item.SetMetadata(MetadataKeys.HintPath, item.ItemSpec);
+        private void SetImplicitMetadataForFrameworkAssemblies()
+        {
+            foreach (var item in FrameworkAssemblies)
+            {
+                item.SetMetadata(MetadataKeys.NuGetIsFrameworkReference, "true");
+                item.SetMetadata(MetadataKeys.Pack, "false");
+                item.SetMetadata(MetadataKeys.Private, "false");
+            }
+        }
 
-                    if (MarkPackageReferencesAsExternallyResolved)
-                    {
-                        item.SetMetadata(MetadataKeys.ExternallyResolved, "true");
-                    }
-                });
+        private void LogMessagesToMSBuild()
+        {
+            if (!EmitAssetsLogMessages)
+            {
+                return;
+            }
 
-            if (!DisableFrameworkAssemblies)
+            foreach (var item in LogMessages)
             {
-                FrameworkAssemblies = RaiseFrameworkAssemblies(
-                    compileTimeTarget,
-                    setup: (asset, item) =>
-                     {
-                         item.SetMetadata(MetadataKeys.Pack, "false");
-                         item.SetMetadata(MetadataKeys.Private, "false");
-                         item.SetMetadata(MetadataKeys.NuGetIsFrameworkReference, "true");
-                     });
+                string message = item.ItemSpec;
+                string severity = item.GetMetadata(MetadataKeys.Severity);
+                string code = item.GetMetadata(MetadataKeys.DiagnosticCode);
+
+                switch (severity)
+                {
+                    case nameof(LogLevel.Error):
+                        Log.LogError(null, code, null, ProjectPath, 0, 0, 0, 0, message);
+                        break;
+                    case nameof(LogLevel.Warning):
+                        Log.LogWarning(null, code, null, ProjectPath, 0, 0, 0, 0, message);
+                        break;
+                    default:
+                        Log.LogMessage(null, code, null, ProjectPath, 0, 0, 0, 0, message);
+                        break;
+                }
             }
+        }
 
-            ContentFilesToPreprocess = RaisePackageAssets(
-                runtimeTarget,
-                p => p.ContentFiles,
-                filter: asset => !string.IsNullOrEmpty(asset.PPOutputPath),
-                setup: (asset, item) => 
+        private byte[] HashSettings()
```

Comment:
Do we have any way of helping to keep this method up to date, so that whenever there is a property that would affect the result we make sure to add it to the hash?

Reply:
I've added test coverage for that in 7b0dbb04f63ec9947a11c6e3a8a477e4048ad9ba

There was actually a bug that this fixes: if $(ProjectAssetsFile) or $(ProjectAssetsCacheFile) were to change between builds, we might consider a cache from a different assets file to be useable.

This would also have been to get wrong over time, so the test is definitely valuable. Great feedback!